core/src/test/kotlin/com/exactpro/th2/read/db/impl/grpc/DataBaseReaderGrpcServerIntegrationTest.kt

```diff
@@ -205,11 +211,187 @@ class DataBaseReaderGrpcServerIntegrationTest {
+                { cfg.queries[it] ?: error("'$it' query isn't found in custom config") },
+                onEvent,
+            )
+
+            val expectedData = persons.asSequence().drop(10).take(9).toList()
+            val responses = mutableListOf<QueryResponse>()
+            GrpcTestHolder(service).use { (stub) ->
+                withCancellation {
+                    val iterator: Iterator<QueryResponse> = stub.execute(request)
+                    // Grpc listener executes the checks sequentially if previous is false
+                    //  1) is isReady flag on gRPC observer true
+                    //  2) is OnReady call back called on gRPC observer. This check is blocking
+                    // In the initial state OnReady is called and isReady is set true
+                    // Iterations:
+                    //  1) check isReady, send message
+                    //  2) check isReady and OnReady, send message
+                    //  3) check isReady and OnReady, block
+                    // Operation sequence: send, (send, receive)
+                    verifyBlocking(genericRowListener,  timeout(1_000).times(2)) {
+                        onRow(any(), any())
+                    }
+                    repeat(expectedData.size) { index ->
+                        // minimal value is 2 because gRPC Listener allow to process first two messages without blocking.
+                        var expectRows = max(2, index + 1)
+                        verifyBlocking(
+                            genericRowListener,
+                            timeout(100).times(expectRows)
+                                .description("Check listener before iterator hasNext call for $index person")
+                        ) { onRow(any(), any()) }
+                        assertTrue(iterator.hasNext(), "Check next gRPC response for $index person")
+                        responses.add(iterator.next())
+
+                        // Client requests the next message after getting current.
+                        // minimal value is 2 because gRPC Listener allow to process first two messages without blocking.
+                        expectRows = max(2, index + 2)
+                        verifyBlocking(
+                            genericRowListener,
+
+                            timeout(100).times(min(expectRows, expectedData.size))
+                                .description("Check listener after iterator hasNext call for $index person")
+                        ) { onRow(any(), any()) }
+
+                        if (expectRows < expectedData.size) {
+                            verify(onEvent, never()).accept(any(), anyOrNull())
+                        } else if (expectRows > expectedData.size) {
+                            verify(onEvent).accept(any(), anyOrNull())
+                        } else {
+                            verify(onEvent, timeout(100)).accept(any(), anyOrNull())
+                        }
```

Comment:
note (non-blocking): I don't quite understand these if statements... Why do we need them in the test? The test must be linear otherwise it is not clear what we test...

Reply:
This test isn't trivial and I spent some times for implementation, but I would like to save it to check that the whole pipeline works with slow consumer speed.